### PR TITLE
apiserver: return 4xx for invalid json patch

### PR DESCRIPTION
**Pull request by nikhita**
Fixes https://github.com/kubernetes/kubernetes/issues/68224

The apiserver should return 400 when the json patch request contains invalid or "missing" fields and 422 when it returns an error which contains `Testing value %s failed` - this means that the value is not in a state that can be processed by the apiserver.

The error messages are checked against the errors here: https://github.com/kubernetes/kubernetes/blob/master/vendor/github.com/evanphx/json-patch/patch.go

https://github.com/kubernetes/kubernetes/pull/54477 was a similar PR for strategic merge patch.

/cc @liggitt @sttts @CaoShuFeng @caesarxuchao 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by nikhita** [comment]
After this,

```json
I0919 20:56:08.304159   10568 helpers.go:201] server response object: [{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "jsonpatch add operation does not apply: doc is missing path: \"invalid\"",
  "reason": "BadRequest",
  "code": 400
}]
```

**Comment by nikhita** [comment]
cc @felixfbecker

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 1]
```diff
@@ -439,3 +441,15 @@ func interpretPatchError(err error) error {
...
 	}
 }
+
+// interpretJSONPatchError interprets the error and returns an error with appropriate HTTP code.
+func interpretJSONPatchError(err error) error {
+	switch {
+	case strings.Contains(err.Error(), "doc is missing"):
+		return errors.NewBadRequest(err.Error())
+	case strings.Contains(err.Error(), "Testing value") && strings.Contains(err.Error(), "failed"):
+		return errors.NewGenericServerResponse(http.StatusUnprocessableEntity, "", schema.GroupResource{}, "", err.Error(), 0, false)
```
Comment (nikhita) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: maybe we could return a more meaningful error here?

Comment (felixfbecker) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: I would suggest 409 Conflict. From _PATCH Method for HTTP_, 2.2 _Error Handling_ https://tools.ietf.org/html/rfc5789#section-2.2

> Conflicting state: Can be specified with a 409 (Conflict) status
> code when the request cannot be applied given the state of the
> resource. For example, if the client attempted to apply a
> structural modification and the structures assumed to exist did
> not exist (with XML, a patch might specify changing element 'foo'
> to element 'bar' but element 'foo' might not exist).

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 2]
```diff
@@ -439,3 +441,15 @@ func interpretPatchError(err error) error {
 		return err
 	}
 }
+
+// interpretJSONPatchError interprets the error and returns an error with appropriate HTTP code.
+func interpretJSONPatchError(err error) error {
+	switch {
+	case strings.Contains(err.Error(), "doc is missing"):
+		return errors.NewBadRequest(err.Error())
+	case strings.Contains(err.Error(), "Testing value") && strings.Contains(err.Error(), "failed"):
```
Comment (felixfbecker) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: I assume the library does not return specific error types for this?

**Comment by felixfbecker** [comment]
Does this also fix https://github.com/kubernetes/kubernetes/issues/68202 ?

**Comment by nikhita** [comment]
> Does this also fix #68202 ?

Yes and no. The error message needs to be improved:

With `--v=9`:

```
I0919 21:29:48.264147   22750 request.go:942] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server rejected our request due to an error in our request","reason":"Invalid","details":{},"code":422}
F0919 21:29:48.264266   22750 helpers.go:119] The  "" is invalid
```

I just saw https://github.com/kubernetes/kubernetes/pull/68346.
@CaoShuFeng Let me know if I should close this PR.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go (hunk 2 above)
Comment (nikhita) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: > I assume the library does not return specific error types for this?

Unfortunately, no :(

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: I'm not in favor of string text matching for error recognition

**Comment by nikhita** [comment]
https://github.com/kubernetes/kubernetes/pull/68346 looks like a step in the right direction. :+1: 

Closing this PR in favor of it.

/close

**Pull request closed without merging**